Document how to get around storage-limitations that some installations may encounter (#59)
plus3it/nessus-agent-formula: docs/StorageUsage-Linux.md

# Configurations Susceptible to the Problem

This problem only afflicts:

* Single-partition systems where the `/` filesystem has less than 2GiB of free storage prior to the initiation of scanning
* Partitioned systems where the `/opt` filesystem &mdash; or, more typically, an `/opt`-hosting `/` filesystem &mdash; has less than 2GiB of free storage prior to the initiation of scanning

The former is typical of EC2s launched from community Enterprise Linux AMIs or similar images published via the AWS Marketplace. Similar is assumed would happen within other template-driven deployments, particular in public CSPs (Azure, GCP, etc.) but has not been directly observed by this project's maintainers.

# Suggested Fixes/Workarounds

## Single-partition Systems

With typical, cloud-hosted OS image-templates, the `cloud-init` utility is installed. If present, the fix is as simple as increasing the size of the boot-volume by at least 2GiB. The `cloud-init` utility will typically just automatically expand the `/` filesystem onto the additional storage.

> For systems without the `cloud-init` utility installed, it will be necessary to update provisioning processes (e.g. Ansible-based system-configuration) to grow the `/`-hosting partition and expand the `/` filesystem to match the resized partition. Typically, this requires using some combination of the `growpart` utility and the filesystem-utilities for the `/` filesystem's fs-type. Most commonly, the `/` filesystem on Enterprise Linux systems is XFS. If this is the case for your Enterprise Linux operating system, that will mean the use of the `xfs_growfs` utility. Usage of the relevant utilities is outside the scope of this document

Reviewer: ```suggestion
For systems without the `cloud-init` utility installed, it will be necessary to update provisioning processes (e.g. Ansible-based system-configuration) to grow the `/`-hosting partition and expand the `/` filesystem to match the resized partition. Typically, this requires using some combination of the `growpart` utility and the filesystem-utilities for the `/` filesystem's fs-type. Most commonly, the `/` filesystem on Enterprise Linux systems is XFS. If this is the case for your Enterprise Linux operating system, that will mean the use of the `xfs_growfs` utility. Usage of the relevant utilities is outside the scope of this document.
```

## Multi-Partition Systems

> The most-frequently observed method for implementing multiple filesystem on Enterprise Linux hosts is via LVM2. The following guidance assumes the use of LVM2 for the paritioning of the boot-disk and that the `/` filesystem is hosted on an LVM2 volume.

Reviewer: ```suggestion
The most-frequently observed method for implementing multiple filesystems on Enterprise Linux hosts is via LVM2. The following guidance assumes the use of LVM2 for the partitioning of the boot-disk and that the `/` filesystem is hosted on an LVM2 volume.
```

Fixing this configuration requires knowing:

* The device-path of the block-device hosting the LVM2 volume-group
* The name of the LVM2 volume-group hosting the root volumes and filesystems
* The name of logical volume hosting the `filesystem`.

While it's not _strictly_ necessary to know these values, _not_ knowing them considerably increases the complexity of the necessary userData payload[^1].

With the `cloud-init` utility present in the host operating system, the problem can be prevented with a mixed-MIME userData payload. Adjust the following userData payload to reflect your host's _actual_ LVM2 volume-group name, name of the volume hosting the `/` parition and the device-path of the block-device hosting the root filesystem(s):

```
Content-Type: multipart/mixed; boundary="===============BOUNDARY=="
MIME-Version: 1.0
Number-Attachments: 2

--===============BOUNDARY==
Content-Disposition: attachment; filename="cloud.cfg"
Content-Transfer-Encoding: 7bit
Content-Type: text/cloud-config
Mime-Version: 1.0

#cloud-config

# Ensure that we use any available extra storage
growpart:
  mode: auto

>   devices: [
>       '/dev/nvme0n1p4'
>     ]

Reviewer: it feels cleaner to keep this all as yaml, rather than mix yaml and json syntaxes

```suggestion
  devices:
    - /dev/nvme0n1p4
```

--===============BOUNDARY==
MIME-Version: 1.0
Content-Type: text/x-shellscript
Content-Disposition: attachment; filename="00_bootdev_setup.sh"
#!/bin/bash
set -euo pipefail
#

> # UserData script to set up SSM and console-based access

Reviewer: ```suggestion
# UserData script to extend the LVM root volume
```

#
#################################################################

lvextend -rl +100%FREE RootVG/rootVol

--===============BOUNDARY==--
```

The above userData will

1. Attempt to grow the disk-partition known to the system as `/dev/nvme0n1p4`.

> 2. Attempt to grow the LVM2 volume-object (that lives on the `/dev/nvme0n1p4` blocke-device node)

Reviewer: ```suggestion
2. Attempt to grow the LVM2 volume-object (that lives on the `/dev/nvme0n1p4` block-device node).
```

If multipart-MIME[^2] is undesirable[^3], the `cloud-config` block can be removed and its logic reimplemented like:

```
growpart /dev/nvme0n1 4
```

In the BASH block _before_ the `lvextend` operation. The multipart-MIME is somewhat more-appropriate when pursuing a more-flexible and/or modular approach to the problem.

### `/opt` Hosted On A Standalone Filesystem

> This solution is basically the same as the "`/opt` Hosted Within the `/` Filesystem" solution. However, it assumes that the `/opt` filesystem already exists as a separate filesystem and associated LVM2 volume. Substitute valuse from the `/` filesystem solution as appropriate.

Reviewer: ```suggestion
This solution is basically the same as the "`/opt` Hosted Within the `/` Filesystem" solution. However, it assumes that the `/opt` filesystem already exists as a separate filesystem and associated LVM2 volume. Substitute values from the `/` filesystem solution as appropriate.
```

### Moving `/opt/nessus_agent` To Its Own filesystem

> This section assumes that the boot-disk has already been expanded and the parition hosting the LVM2 volume-group has been grown.

Reviewer: ```suggestion
This section assumes that the boot-disk has already been expanded and the partition hosting the LVM2 volume-group has been grown.
```

> While the entire `/opt` filesystem _could_ be divorced from the `/` filesystem, the `/opt` filesystem typically already has content within it. Further, that content oftent results in open file-handles that make launch-time automation more problematic to account for. To ensure that _just_ the `/opt/nessus_agent` content ends up on a dedicated partition, userData payload similar to:

Reviewer: ```suggestion
While the entire `/opt` filesystem _could_ be divorced from the `/` filesystem, the `/opt` filesystem typically already has content within it. Further, that content often results in open file-handles that make launch-time automation more problematic to account for. To ensure that _just_ the `/opt/nessus_agent` content ends up on a dedicated partition, userData payload similar to:
```

```
[...ELIDED...]

--===============BOUNDARY==
MIME-Version: 1.0
Content-Type: text/x-shellscript
Content-Disposition: attachment; filename="01_nessusdev_setup.sh"
#!/bin/bash
set -euo pipefail
#

> # UserData script to set up SSM and console-based access

Reviewer: ```suggestion
# UserData script to create a dedicated LVM volume mount for the Nessus agent
```
